typescript/core/src/ChunkCursor.ts

```diff
@@ -211,80 +150,78 @@ export class ChunkCursor {
-      if (reverse) {
-        if (this.#endTime != undefined) {
-          startIndex = sortedIndexBy(result.record.records, this.#endTime, (logTime) => -logTime);
-        }
-      } else {
-        if (this.#startTime != undefined) {
-          startIndex = sortedIndexBy(result.record.records, this.#startTime, (logTime) => logTime);
-        }
-      }
-
-      if (startIndex >= result.record.records.length) {
-        continue;
-      }
-      if (reverse) {
-        if (
-          this.#startTime != undefined &&
-          result.record.records[startIndex]![0] < this.#startTime
-        ) {
-          continue;
-        }
-      } else {
-        if (this.#endTime != undefined && result.record.records[startIndex]![0] > this.#endTime) {
-          continue;
-        }
-      }
-
-      this.#messageIndexCursors.push({
-        index: startIndex,
-        channelId: result.record.channelId,
-        records: result.record.records,
-      });
+      this.#orderedMessageOffset = this.#orderedMessageOffset.concat(result.record.records);
     }
 
     if (offset !== messageIndexesView.byteLength) {
       throw new Error(
         `${messageIndexesView.byteLength - offset} bytes remaining in message index section`,
       );
     }
-  }
 
-  #getSortTime(): bigint {
-    if (!this.#messageIndexCursors) {
-      return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
+    if (this.#orderedMessageOffset.length === 0) {
+      return;
     }
 
-    const cursor = this.#messageIndexCursors.peek();
-    if (!cursor) {
+    this.#orderedMessageOffset.sort(([logTimeA], [logTimeB]) => Number(logTimeA - logTimeB));
+    if (reverse) {
+      // If we used `logTimeB - logTimeA` as the comparator for reverse iteration, messages with
+      // the same timestamp would not be in reverse order. To avoid this problem we use reverse()
+      // instead.
+      this.#orderedMessageOffset.reverse();
+    }
+
+    const [logTimeFirstMessage] = this.#orderedMessageOffset[0]!;
+    if (logTimeFirstMessage < this.chunkIndex.messageStartTime) {
       throw new Error(
-        `Unexpected empty cursor for chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+        `Chunk contains a message with logTime (${logTimeFirstMessage}) earlier than chunk messageStartTime (${this.chunkIndex.messageStartTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
       );
     }
 
-    return cursor.records[cursor.index]![0];
+    const [logTimeLastMessage] = this.#orderedMessageOffset[this.#orderedMessageOffset.length - 1]!;
+    if (logTimeLastMessage > this.chunkIndex.messageEndTime) {
+      throw new Error(
+        `Chunk contains a message with logTime with logTime (${logTimeLastMessage}) later than chunk messageEndTime (${this.chunkIndex.messageEndTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+      );
+    }
+
+    // Remove message offsets whose log time is not within [startTime, endTime];
```

Comment:
To determine the ChunkCursor's first mesage logTime which is returned in `#getSortTime` and used for ordering ChunkCursors in the heap

Reply:
I changed this part now a bit to make it clearer. Instead of pruning the message offset list from unused offsets (which are not in the range of [startTime, endTime]) I only determine the start & endIndex which might also be a bit more performant.